### DISCUSSION TITLE
Headless UI's `Listbox` does not trigger a form's `onChange` when using React Hook Form's `Controller` (but native `select` does)

### DISCUSSION THREAD

**Discussion by bennettdams** [react-hook-form/react-hook-form#9359] [Q&A] [answered]
I want to trigger a form submit every time a selection is made.

Using Headless UI's `Listbox` together with React Hook Form's `Controller` does not fire the form's `onChange`.

Using a native `select` with the same setup does fire the `onChange` though, that's why I created the issue in the Headless UI repo, but I'm not sure, so maybe someone already encountered this? See: https://github.com/tailwindlabs/headlessui/issues/2003

Reproduction:

https://codesandbox.io/s/rhf-onchange-with-headlessuis-listbox-ce917d?file=/src/App.tsx

In my reproduction you can see that the form is supposed to submit when the selection changes. Here, it logs to console, e.g.:

![image](https://user-images.githubusercontent.com/29319414/200953926-e1e0a209-6235-4da6-bbe6-e83fd9dfb1d9.png)

In the UI there are two selects: The first is via `Listbox`, the second is with a native `select`.
When selecting an entry via `Listbox`, the value changes, but the form is not submitted (unexpected).
When selecting an entry via the native `select`, the value changes and the form is submitted (as expected).

Code for reference, see CodeSandbox for full code:

```tsx
function SelectNative({
  selected,
  onChange,
  name
}: {
  selected: string;
  onChange: (newSelected: string) => void;
  name: string;
}) {
  return (
    <select
      value={selected}
      onChange={(newSelected) => onChange(newSelected.currentTarget.value)}
      name={name}
    >
      {people.map((person) => (
        <option key={person.id} value={person.id}>
          {person.name}
        </option>
      ))}
    </select>
  );
}

      <p>With native select</p>
      <form onChange={handleSubmit(onSubmit)}>
        <Controller
          control={control}
          name="selectedPersonId"
          render={({ field }) => (
            <SelectNative
              selected={field.value}
              onChange={(sel) => field.onChange(sel)}
              name={field.name}
            />
          )}
        />
      </form>



function SelectWithListbox({
  selected,
  onChange,
  name
}: {
  selected: string;
  onChange: (newSelected: string) => void;
  name: string;
}) {
  return (
    <Listbox value={selected} onChange={onChange} name={name}>
      <Listbox.Button>
        {people.find((person) => person.id === selected)?.name}
      </Listbox.Button>
      <Listbox.Options>
        {people.map((person) => (
          <Listbox.Option key={person.id} value={person.id}>
            {person.name}
          </Listbox.Option>
        ))}
      </Listbox.Options>
    </Listbox>
  );
}

    <div>
      <p>With Listbox</p>
      <form onChange={handleSubmit(onSubmit)}>
        <Controller
          control={control}
          name="selectedPersonId"
          render={({ field }) => (
            <SelectWithListbox
              selected={field.value}
              onChange={(sel) => field.onChange(sel)}
              name={field.name}
            />
          )}
        />
      </form>
```

**Comment by bluebill1049** [comment]
could you double-check if `field.onChange` is get invoked? useController/Controller get value updated via it gets executed.

**Reply by bennettdams** [reply]
I added some `console.log`s to the CodeSandbox above, both variants have their `onChange` callback invoked and `field.onChange` is called right after the `console.log` - or do you mean something else?

![image](https://user-images.githubusercontent.com/29319414/201449670-098d6296-cb48-409f-9172-5d56fadf825e.png)

**Comment by leapful** [comment, marked as answer]
@bennettdams @bluebill1049 

The ListBox component from Headless UI is not native `select` element therefore when changing its selected value, event "onClick" is fired, not "onChange" event and there was no event propagation up to parent element `form` too.

A workaround might be considered by using custom dispatch event.

Working demo: https://codesandbox.io/s/rhf-onchange-with-headlessuis-listbox-forked-vmuslf?file=/src/App.tsx

**Reply by bluebill1049** [reply]
interesting! thanks for looking into this.

**Reply by bennettdams** [reply]
Thanks, @leapful! I will add your solution to [my issue in the Headless UI repo](https://github.com/tailwindlabs/headlessui/issues/2003) and will ask whether this can (or even should) be part of the `Listbox` component directly.
